urllib3/connectionpool.py

```diff
@@ -568,25 +568,22 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
             # Close the connection. If a connection is reused on which there
             # was a Certificate error, the next request will certainly raise
             # another Certificate error.
-            if conn:
-                conn.close()
-                conn = None
+            conn = conn and conn.close()
+            release_conn = True
```

Comment:
In the case of BaseSSLError, CertificateError or SSLError you raise right after setting release_conn to True but where is that exception caught? - I think that it will miss the finally block where you do that actual release with self._put_conn so it looks like those connections will never make it back to the pool.

In the TimeoutError, HTTPException, SocketError, ConnectionError exception handler you do not raise an exception. You set release_conn to True, release it in the finally block and then make a recursive call to urlopen, this time with release_conn set to True. If that succeeds you'll then release the conn back to the pool (because release_conn is now true) before the response is read, meaning that it's possible that when response.read() happens the connection might be in use by someone else.

Apologies if I'm wrong about this cos I haven't tested it yet but just by inspecting the code it looks that way.